### PR TITLE
fix: don't let `loading="lazy"` overwrite a working image `src`

### LINKED ISSUE #218: fix_lazy_images overwrites a valid `src` on any `img[loading="lazy"]`
Thanks for the crate — it is the extraction engine behind a reading browser
I am building, and in a recent comparison of 13 Rust HTML-extraction crates it
was the only readability port that reliably found the main content of a page.

I believe `fix_lazy_images` deviates from readability.js in a way that damages
real pages.

## What happens

`MATCHER_LAZY_IMG` is `[class*="lazy"], img[loading="lazy"]` (`src/glob.rs`), so
in `fix_lazy_images` (`src/prep_article.rs`) this early-continue never fires for
an ``, even when its `src` is a perfectly good URL:

```rust
if (node.has_attr("src") || node.has_attr("srcset")) && !node.is_match(&MATCHER_LAZY_IMG) {
    continue;
}
```

The attribute scan below then copies the first attribute whose value contains an
image extension into `src`, overwriting the working one.

In readability.js the corresponding guard looks at the class name only, so an
image that already has a `src` is left alone — `loading="lazy"` is not treated
as a lazy-loading marker:

```js
if ((elem.src || (elem.srcset && elem.srcset != "null")) &&
    !elem.className.toLowerCase().includes("lazy")) {
  return;
}
```

## Why it matters in practice

Wikipedia puts an RDFa `resource` attribute on every image, pointing at the
*file description page*:

```html
<img resource="https://en.wikipedia.org/wiki/File:Foo.jpg"
     src="//upload.wikimedia.org/wikipedia/commons/9/9d/Foo.jpg"
     loading="lazy" class="mw-file-element">
```

`resource` ends in `.jpg`, so it wins, and the extracted article points at an
HTML page instead of an image — the reader gets ~140 KB of HTML where a photo
was expected. On `en.wikipedia.org/wiki/Rust_(programming_language)` 5 of the 11
images are affected; on ru.wikipedia biography articles it is usually the lead
portrait. Images whose `resource` ends in `.svg` survive by accident, because
`.svg` is not in `IMG_EXT`.

Since `loading="lazy"` is present on most images on the modern web, any
attribute that happens to look like an image filename (`resource`,
`data-file-name`, `title`, …) can replace a working `src`.

## Minimal reproduction

dom_smoothie 0.18.1:

```rust
use dom_smoothie::{Config, Readability};

const TEXT: &str = "Kenneth Arrow proved the impossibility theorem for collective choice, \
and it reshaped welfare economics. The paragraphs below exist only so that the readability \
algorithm has enough text to pick this node as the article body of the document.";

fn main() {
    let html = format!(
        "<html><body><article><h1>Arrow</h1>\
         <p><img loading=\"lazy\" src=\"https://upload.example.org/commons/portrait.jpg\" \
         resource=\"https://example.org/wiki/File:Portrait.jpg\"></p>\
         <p>{TEXT}</p><p>{TEXT}</p><p>{TEXT}</p></article></body></html>"
    );

    let mut r = Readability::new(
        html.as_str(),
        Some("https://example.org/wiki/Arrow"),
        Some(Config::default()),
    ).unwrap();
    let article = r.parse().unwrap();

    let content = article.content.to_string();
    let src = content.split("src=\"").nth(1)
        .and_then(|s| s.split('"').next())
        .unwrap_or("<no img>");
    println!("src in article: {src}");
}
```

Prints:

```
src in article: https://example.org/wiki/File:Portrait.jpg
```

Expected: `https://upload.example.org/commons/portrait.jpg`.

## Suggestion

Treat `loading="lazy"` as a lazy marker only when there is no usable `src` — the
base64-placeholder branch above already removes such a `src`, so the attribute
scan still runs for genuinely lazy images. In other words, keep the
early-continue when `src` is present and is not a `data:` URL, regardless of
`loading`.

As a workaround I currently remove the `loading` attribute from images that
already have a non-`data:` `src` before handing the document to
`Readability::with_document`, which restores readability.js behaviour. Happy to
send a PR if you would like this in the crate.

### PR DESCRIPTION
**Pull request by gurov**
`fix_lazy_images` treated every `img[loading="lazy"]` as a lazy-loading placeholder and scanned the rest of the attributes for something that looks like an image path. `loading` is a rendering hint, though: it says nothing about the address, and pages put it on images whose `src` is perfectly good. Wikipedia is the visible case -- its `resource` attribute points at the file description page, so 5 of 12 images of the "Rust (programming language)" article ended up with `.../wiki/File:Name.png` instead of the image itself, which is an html page of ~140 KB where the reader expects a picture.

`Readability.js` enables the attribute scan only for a `lazy` class, and that part stays as it is. The `loading` hint is kept as well, but now it counts only while the node has no usable address of its own, so markup that really does hold a placeholder (a `data:` stub with the real address in another attribute) is still repaired.

Closes #218

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed lazy-loaded image handling so images with `loading="lazy"` retain a working `src` or `srcset`.
  - Prevented lazy-loading hints from overwriting valid image addresses.
  - Improved recognition of placeholder data URLs when determining whether an image has a usable address.
  - Improved handling of images with empty or whitespace-only source attributes.

- **Documentation**
  - Added an Unreleased changelog entry describing the corrected lazy-image behavior.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,13 @@
 
 All notable changes to the `dom_smoothie` crate will be documented in this file.
 
+## [Unreleased]
+
+### Fixed
+- `fix_lazy_images` no longer replaces a working `src` on images marked with `loading="lazy"`.
+The hint is now taken into account only while the image has no usable address of its own,
+which is closer to `Readability.js`, where only a `lazy` class enables the attribute scan (#218).
+
```
Comment (niklak) [inline CHANGELOG.md]: - `fix_lazy_images` no longer replaces a working `src` on images marked with `loading="lazy"` (by @gurov in #218).

**Comment by niklak** [comment]
@gurov, sorry for the review delay, I'm a bit busy right now.

Could you make the comments more compact? This is internal code anyway.
Could you combine these into a single table test instead of three? It’ll make adding new test cases (like preceding whitespace) much easier.

File: src/matching.rs [hunk 2]
```diff
@@ -163,8 +163,27 @@ pub(crate) fn contains_share_elements(value: &str) -> bool {
...
+/// and keeps the real address in another attribute.
+pub(crate) fn has_image_address(node: &NodeRef) -> bool {
+    ["src", "srcset"].iter().any(|name| {
+        node.attr(name).is_some_and(|value| {
+            let value = value.trim();
+            !value.is_empty() && !is_data_url(value)
+        })
+    })
+}
+
```
Comment (niklak) [inline src/matching.rs:176]:
```rust
pub(crate) fn has_image_address(node: &NodeRef) -> bool {
    node.attrs().iter().any(|attr| {
        matches!(attr.name.local.as_ref(), "src" | "srcset")
            && {
                let value = attr.value.trim();
                !value.is_empty() && !is_data_url(value)
            }
    })
}
```

**Comment by gurov** [comment]
> @gurov, sorry for the review delay, I'm a bit busy right now.
> 
> Could you make the comments more compact? This is internal code anyway. Could you combine these into a single table test instead of three? It’ll make adding new test cases (like preceding whitespace) much easier.

Done! `has_image_address` now walks `attrs()` once (your code), the comments are reduced, and the three tests are one table; the whitespace-only src case you mentioned is in it. CHANGELOG uses your wording.

**Comment by niklak** [comment]
@gurov, thank you!
